sw/host/opentitanlib/src/transport/hyperdebug/mod.rs

```diff
@@ -0,0 +1,320 @@
+                                interface: interface.number(),
+                                in_endpoint,
+                                out_endpoint,
+                            });
+                        }
+                    }
+                }
+            }
+        }
+        let mut spi_names = HashMap::new();
+        // Eventually, the SPI aliases below should either go into configuration file, or come
+        // from the HyperDebug firmware, declaring what it supports (as is the case with UARTs.)
+        spi_names.insert("SPI2".to_string(), 0);
+        spi_names.insert("0".to_string(), 0);
+        let result = Hyperdebug {
+            spi_names,
+            spi_interface: spi_interface.ok_or(Error::CommunicationError("Missing SPI interface"))?,
+            uart_ttys,
+            inner: Rc::new(Inner {
+                console_tty: console_tty.ok_or(Error::CommunicationError("Missing console interface"))?,
+                usb_device: RefCell::new(device),
+                gpio: Default::default(),
+                spis: Default::default(),
+                uarts: Default::default(),
+            })
+        };
+        Ok(result)
+    }
+
+    /// Locates the /dev/ttyUSBn node corresponding to a given interface in the sys directory
+    /// tree, e.g. /sys/bus/usb/devices/1-4/1-4:1.0 .
+    fn find_tty(path: &Path) -> Result<PathBuf> {
+        for entry in fs::read_dir(path)? {
+            let dir_entry = entry?;
+            match dir_entry.file_name().into_string() {
+                Ok(filename) => {
+                    if filename.starts_with("tty") {
+                        return Ok(PathBuf::from("/dev").join(dir_entry.file_name()));
+                    }
+                }
+                _ => ()
+            }
+        }
+        Err(Error::CommunicationError("Did not find ttyUSBn device").into())
+    }
+}
+
+impl Inner {
+
+    /// Send a command to HyperDebug firmware, with a callback to receive any output.
+    pub fn execute_command(
+        &self,
+        cmd: &str,
+        mut callback: impl FnMut(&str)) -> Result<()>
+    {
+        use std::io::Read;
+        use std::io::Write;
+        let mut port = serialport::new(self.console_tty.to_str().ok_or(Error::UnicodePathError)?, 115_200)
+            .timeout(std::time::Duration::from_millis(100))
+            .open().expect("Failed to open port");
+
+        // Ideally, we would invoke Linux flock() on the serial
+        // device, to detect minicom or another instance of
+        // opentitantool having the same serial port open.  Incoming
+        // serial data could go silenly missing, in such cases.
+        let mut buf = [0u8; 128];
+        loop {
+            match port.read(&mut buf[0..128]) {
+                Ok(rc) => {
+                    log::info!("Discarded {} characters: {:?}\n", rc,
+                               &std::str::from_utf8(&buf[0..rc]));
+                }
+                Err(error) => {
```

Comment:
I think it would be more helpful to write guards into this case so as not to indent further:
```rust
Err(error) if error.kind() == TimedOut => break,
Err(error) => ...,
```

Reply:
Good point, done.